File: php-laravel-equitab/app/Models/Transaction.php
<?php

namespace App\Models;

use App\Observers\TransactionObserver;
use Illuminate\Database\Eloquent\Attributes\ObservedBy;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

#[ObservedBy([TransactionObserver::class])]
class Transaction extends Model
{
    /** @use HasFactory<\Database\Factories\TransactionFactory> */
    use HasFactory;

    protected $fillable = [
        'name',
        'cost',
        'location',
        'datetime',
        'category_id',
        'payer',
        'owers'
    ];

    protected $with = [
        'payer',
        'owers',
        'category',
    ];

    protected $casts = [
        'cost' => 'float',
        'datetime' => 'datetime:c'
    ];

    public function setPayerAttribute(array $payer)
    {
        $this->payer_id = $payer['id'];
    }

    public function setOwersAttribute(array $owers)
    {
        if ($this->id) {
            $this->owers()->sync(collect($owers)->mapWithKeys(fn($o) => [$o['id'] => ['aggregate' => $o['aggregate']]])->toArray());
        }
    }

    public function payer()
    {
        return $this->belongsTo(User::class);
    }

    public function owers()
    {
        return $this->belongsToMany(User::class, TransactionOwer::class, 'transaction_id', 'ower_id')
            ->withPivot('aggregate');
    }

    public function ledger()
    {
        return $this->belongsTo(Ledger::class);
    }

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function products()
    {
        return $this->hasMany(Product::class)->orderBy('index');
    }
}

File: swift-ios-equitab/equitab/Models/Transaction.swift
import Foundation
import SwiftData

@Model
final class Transaction: Codable {
    @Attribute(.unique) var id: Int
    var name: String
    var cost: Double
    var location: String
    var datetime: Date
    var category: Category
    var payer: User
    var owers: [User]
    var ledger: Ledger?
    var products: [Product]?
    var productCount: Int?

    private enum CodingKeys: String, CodingKey {
        case id, name, cost, location, datetime, category, payer, owers, ledger, products
        case productCount = "product_count"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        cost = try container.decode(Double.self, forKey: .cost)
        location = try container.decode(String.self, forKey: .location)
        datetime = try container.decode(Date.self, forKey: .datetime)
        category = try container.decode(Category.self, forKey: .category)
        payer = try container.decode(User.self, forKey: .payer)
        owers = try container.decode([User].self, forKey: .owers)
        ledger = try container.decodeIfPresent(Ledger.self, forKey: .ledger)
        products = try container.decodeIfPresent([Product].self, forKey: .products)
        productCount = try container.decodeIfPresent(Int.self, forKey: .productCount)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(cost, forKey: .cost)
        try container.encode(location, forKey: .location)
        try container.encode(datetime, forKey: .datetime)
        try container.encode(category, forKey: .category)
        try container.encode(payer, forKey: .payer)
        try container.encode(owers, forKey: .owers)
        try container.encode(ledger, forKey: .ledger)
        try container.encode(products, forKey: .products)
        try container.encode(productCount, forKey: .productCount)
    }
}
